src/components/main/mod.rs

```diff
@@ -24,38 +26,95 @@ pub struct Prop {
     };
-    use_future(&cx, (), |_| async move {
-        loop {
-            if let Ok(list) = rg.list_conversations().await {
-                let mut current_conversations: HashMap<Uuid, ConversationInfo> = HashMap::new();
-                for item in &list {
-                    let to_insert = match st.read().all_chats.get(&item.id()) {
-                        Some(v) => v.clone(),
-                        None => ConversationInfo {
-                            conversation: item.clone(),
-                            creation_time: DateTime::from(Local::now()),
-                            ..Default::default()
-                        },
-                    };
-                    current_conversations.insert(item.id(), to_insert);
+
+    use_future(&cx, &rg, |mut rg| async move {
+        log::debug!("streaming conversations");
+
+        // todo: only accept incoming conversations from people we are friends with.
+
+        // get all conversations and update state
+        let mut conversations: HashMap<Uuid, Conversation> = HashMap::new();
+        match rg.list_conversations().await {
+            Ok(r) => {
+                for c in r {
+                    conversations.insert(c.id(), c.clone());
+                }
+            }
+            Err(e) => {
+                log::error!("failed to get conversations: {}", e);
+                return;
+            }
+        }
+
+        // detect removed conversations
+        for id in state.read().all_chats.keys() {
+            if !conversations.contains_key(id) {
+                log::debug!("removing chat");
+                state.write().dispatch(Actions::HideChat(*id));
+            }
+        }
+
+        // detect added conversations
+        for (id, conv) in conversations {
+            if !state.read().all_chats.contains_key(&id) {
+                log::debug!("adding chat");
+                state
+                    .write()
+                    .dispatch(Actions::AddConversation(conv.clone()));
+            }
+        }
+
+        // receive events from Warp
+        let mut stream = loop {
+            match rg.subscribe().await {
+                Ok(stream) => break stream,
+                Err(warp::error::Error::MultiPassExtensionUnavailable)
+                | Err(warp::error::Error::RayGunExtensionUnavailable) => {
+                    tokio::time::sleep(std::time::Duration::from_millis(500)).await;
+                }
+                Err(_e) => {
+                    //Should not reach this point but should handle an error if it does
+                    tokio::time::sleep(std::time::Duration::from_millis(500)).await;
+                }
+            }
+        };
+
+        while let Some(event) = stream.next().await {
+            match event {
+                RayGunEventKind::ConversationCreated { conversation_id } => {
+                    //For now get the conversation from list_conversation
+                    let conversation = rg
```

Comment:
This can be a `if let Some(...)` or maybe use a match condition (although https://github.com/Satellite-im/Uplink/pull/489 does introduce `RayGun::get_conversation` so this would not be needed)

Reply:
I'm not sure I follow. it looks like `ConversationCreated` and `ConverstionDeleted` both require handling. it looks like I called the wrong event when a conversation is deleted though.

Reply:
I was referring to `let conversation =` part. Though with an update merged in you could do 

```rust
if let Ok(conversation) = rg.get_conversation(conversation_id) {
// code here
}
```